src/util/websocket/WebSocketManager.cpp

```diff
@@ -0,0 +1,241 @@
+#include <ctre/ctre.h>
+
+#include <boost/asio/experimental/awaitable_operators.hpp>
+#include <iostream>
+#include <mutex>
+#include <optional>
+
+#include "../MultiMap.h"
+#include "../http/HttpUtils.h"
+#include "./Common.h"
+#include "./QueryState.h"
+
+namespace ad_utility::websocket {
+
+class WebSocketId {
+  uint32_t id_;
+  explicit WebSocketId(uint32_t id) : id_{id} {}
+
+ public:
+  // Generate unique webSocketIds until it overflows and hopefully the old
+  // ids are gone by then.
+  static WebSocketId uniqueId() {
+    static std::atomic<uint32_t> webSocketIdCounter{0};
+    return WebSocketId(webSocketIdCounter++);
+  }
+
+  constexpr bool operator==(const WebSocketId&) const noexcept = default;
+  constexpr bool operator!=(const WebSocketId&) const noexcept = default;
+
+  friend std::hash<WebSocketId>;
+};
+
+}  // namespace ad_utility::websocket
+
+template <>
+struct std::hash<ad_utility::websocket::WebSocketId> {
+  auto operator()(
+      const ad_utility::websocket::WebSocketId& webSocketId) const noexcept {
+    return std::hash<uint32_t>{}(webSocketId.id_);
+  }
+};
+
+namespace ad_utility::websocket {
+using namespace boost::asio::experimental::awaitable_operators;
+using websocket = beast::websocket::stream<tcp::socket>;
+using common::TimedClientPayload;
+
+using QueryUpdateCallback = std::function<void(TimedClientPayload)>;
+
+static std::mutex activeWebSocketsMutex{};
+static absl::btree_multimap<QueryId, WebSocketId> activeWebSockets{};
+// Note there may be active websockets that are not currently listening
+// because of concurrency.
+static std::mutex listeningWebSocketsMutex{};
+static absl::flat_hash_map<WebSocketId, QueryUpdateCallback>
+    listeningWebSockets{};
+
+void registerCallback(const QueryId& queryId, WebSocketId webSocketId,
+                      QueryUpdateCallback callback) {
+  std::lock_guard lock1{activeWebSocketsMutex};
+  // make sure websocket has not been shut down right before acquiring the lock
+  if (containsKeyValuePair(activeWebSockets, queryId, webSocketId)) {
+    std::lock_guard lock2{listeningWebSocketsMutex};
+    // Ensure last callback has been fired, otherwise co_await will wait
+    // indefinitely
+    AD_CORRECTNESS_CHECK(listeningWebSockets.count(webSocketId) == 0);
+    listeningWebSockets[webSocketId] = std::move(callback);
+  }
+}
+
+// ONLY CALL THIS IF YOU HOLD the listeningWebSocketsMutex!!!
+void fireCallbackAndRemoveIfPresentNoLock(WebSocketId webSocketId,
+                                          TimedClientPayload payload) {
+  if (listeningWebSockets.count(webSocketId) != 0) {
+    listeningWebSockets.at(webSocketId)(std::move(payload));
+    listeningWebSockets.erase(webSocketId);
+  }
+}
+
+void fireCallbackAndRemoveIfPresent(WebSocketId webSocketId,
+                                    TimedClientPayload payload) {
+  std::lock_guard lock{listeningWebSocketsMutex};
+  fireCallbackAndRemoveIfPresentNoLock(webSocketId, std::move(payload));
+}
```

Comment:
The `NoLock` is a reason to make such things classes etc.

Reply:
In theory you could of course move the "no lock" definition to the header so it can be accessed elsewhere, and making the whole thing private would prevent that, but obviously this doesn't solve the issue that you could accidentally call the wrong function or am I missing something?